### DISCUSSION TITLE
Adding new laser device - qudi hw wrapping convention - module.Class ...

### DISCUSSION THREAD

**Discussion by AItah** [Ulm-IQO/qudi-iqo-modules#111] [Q&A] [answered]
Hi,

We would like to add new laser module. We already operate the module using a standalone python script. Next stage to wrap it correctly according to qudi convention - however we are not sure about it.

looking at Coherent obis as an example we see the following:
   
```
Example config for copy-paste:

   obis_laser:
        module.Class: 'laser.coherent_obis_laser.OBISLaser'
        options:
            com_port: 'COM3'
```

can you please explain the name (`obis_laser:`) and module.Class convention?
Does the naming need to be similar to the filename we will create and/or class name, etc?

BR,
Amir

**Comment by Geegee** [comment]
The entry in the config file you can name as you like ("obis_laser" could
be "obis_laser_serial_num123" or what you like.
In essence if my understading is correct qudi needs to find the hardware
class.
Therefore your config entry needs to be under the hardware level i.e.
hardware:
    my_hardware_module:
        module.Class:
'relative.path.to.hardwarefile.from.hardware.folder.NameOfHardwareClass'
        options:
           var: "Internal variable used by the hardware class"

Hope this helps you.
Regards,

Gerhard

Am Do., 9. Nov. 2023 um 13:44 Uhr schrieb AItah ***@***.***>:

**Comment by AItah** [comment, marked as answer]
Thank you. 
So if my class is COBOLTLaser and the file is located under ../HW/laser/cobolt_06_mld_laser.py the following should be the config settings?
```
Example config for copy-paste:

    cobolt_06_mld_laser:
        module.Class: 'laser.cobolt_06_mld_laser.COBOLTLaser
        options:
            port_id: '0'
```

**Comment by gemmell** [comment]
EDIT: You can actually ignore this. I realised you were doing something much simpler than what I thought you were doing :) See my reply :)

It's slightly more complicated than that. 

It'll load the module (and class) with a prefixed "qudi.hardware" module name. So what I suggest is that you make a package like this:

```
my-package\
   src\
      qudi\
         hardware\         
            my_company\
               __init__.py
               cobolt_laser.py
    tests\...
    pyproject.toml
```

The important thing here is that your package contains two namespace modules "qudi" and inside that, "hardware". So now if you `pip install -e .` your package, it'll install it _into the package namespace that qudi-core looks for modules_! Note that they are namespace modules because they _don't_ have an `__init__.py` in them - the `my_company` package contains the `__init__.py` which tells it it's a proper package (not just a namespace).

Then in your config file, you say:
```
my_company_laser:
        module.Class: 'my_company.cobolt_laser.CoboltLaser'
    
```

Of course your `cobolt_laser.py` will need to contain a class called `CoboltLaser` that inherits from `SimpleLaserInterface` (or whatever it is) and does what it needs to do. 

You might get better value out of having the functionality to drive the laser in one class (what you have already), and writing your cobolt_laser.py to contain just the adapter to make it work with Qudi. Better from a software engineering perspective anyway in case you ever need to drive it from other things or software (e.g. notebooks).

**Reply by gemmell** [reply]
Oh, if you're adding it directly to this repo with a mind to push it back (why did I not think you were?), you don't need your own module, you can just jam it in the `hardware` folder & it'll pick it up because that's already in that namespace. And yes, what you wrote seems correct.
